poly_borders/src/borders_data.cpp

```diff
@@ -0,0 +1,521 @@
+  Platform::GetFilesByExt(bordersDir, kBorderExtension, files);
+
+  LOG(LINFO, ("Start read data from .poly files."));
+  for (auto const & file : files)
+  {
+    auto const fullPath = base::JoinPath(bordersDir, file);
+    auto fileCopy = file;
+    size_t id = 1;
+
+    std::vector<m2::RegionD> borders;
+    borders::LoadBorders(fullPath, borders);
+    for (auto const & region : borders)
+    {
+      Polygon polygon;
+
+      m_indexToName[prevIndex] = fileCopy;
+      m_nameToIndex[fileCopy] = prevIndex++;
+      for (auto const & point : region.Data())
+        polygon.m_points.emplace_back(point);
+
+      polygon.m_rect = region.GetRect();
+
+      fileCopy = file + std::to_string(id);
+      ++id;
+
+      m_bordersPolygons.emplace_back(std::move(polygon));
+    }
+  }
+
+  RemoveDuplicatePoints();
+}
+
+void BordersData::MarkPoints()
+{
+  size_t const threadsNumber = std::thread::hardware_concurrency();
+  LOG(LINFO, ("Start marking points, threads number:", threadsNumber));
+
+  base::thread_pool::computational::ThreadPool threadPool(threadsNumber);
+
+  std::vector<std::future<void>> tasks;
+  for (size_t i = 0; i < m_bordersPolygons.size(); ++i)
+  {
+    Processor processor(*this);
+    tasks.emplace_back(threadPool.Submit(processor, i));
+  }
+
+  for (auto & task : tasks)
+    task.wait();
+}
+
+void BordersData::DumpPolyFiles(std::string const & targetDir)
+{
+  size_t n = m_bordersPolygons.size();
+  for (size_t i = 0; i < n; ++i)
+  {
+    if (ShouldLog(i, n))
+      LOG(LINFO, ("Dumping poly files:", i + 1, "/", n));
+
+    auto name = m_indexToName[i];
+    std::vector<std::vector<m2::PointD>> points;
+    size_t id = 1;
+    for (;;)
+    {
+      std::vector<m2::PointD> toAppend;
+      for (auto const & markedPoint : m_bordersPolygons[i].m_points)
+        toAppend.emplace_back(markedPoint.m_point);
+
+      points.emplace_back(std::move(toAppend));
+      if (i + 1 < n && m_indexToName[i + 1] == name + std::to_string(id))
+      {
+        ++i;
+        ++id;
+      }
+      else
+      {
+        break;
+      }
+    }
+
+    std::string const polyPath = base::JoinPath(targetDir, name);
+    DumpPolyFile(polyPath, name, points);
+  }
+}
+
+// static
+void BordersData::DumpPolyFile(std::string const & polyPath, std::string mwmName,
+                               std::vector<std::vector<m2::PointD>> const & polygons)
+{
+  std::ofstream output(polyPath);
+  output << std::setprecision(20);
+
+  static std::string const kPolyExtension = ".poly";
+  mwmName.replace(mwmName.end() - kPolyExtension.size(), mwmName.end(), "");
+
+  output << mwmName << std::endl;
+  for (size_t i = 0; i < polygons.size(); ++i)
+  {
+    output << i + 1 << std::endl;
+    for (auto const & point : polygons[i])
+    {
+      auto const latlon = MercatorBounds::ToLatLon(point);
+      output << latlon.m_lon << " " << latlon.m_lat << std::endl;
+    }
+
+    output << "END" << std::endl;
+  }
+
+  output << "END" << std::endl;
```

Comment:
Can we factor out the code that is common with borders::UnpackBorders and reuse it here?

Reply:
https://github.com/mapsme/omim/pull/11013